connector/google/google.go

```diff
@@ -0,0 +1,192 @@
+// Package google implements logging in through OpenID Connect providers.
+package google
+import (
+	"golang.org/x/oauth2"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"github.com/coreos/go-oidc"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2/google"
+	"github.com/coreos/dex/connector"
+	directory "google.golang.org/api/admin/directory/v1"
+)
+// Config holds configuration options for OpenID Connect logins.
+type Config struct {
+	Issuer       string `json:"issuer"`
+	ClientID     string `json:"clientID"`
+	ClientSecret string `json:"clientSecret"`
+	RedirectURI  string `json:"redirectURI"`
+	Scopes []string `json:"scopes"`
+	// Optional list of whitelisted domains when using Google
+	// If this field is nonempty, only users from a listed domain will be allowed to log in
+	HostedDomains []string `json:"hostedDomains"`
+	ServiceAccountCred  string `json:"serviceAccountCred"`
+	Subject         string `json:"subject"`
+	GroupPrefix	string `json:"groupPrefix"`
+}
+
+// Open returns a connector which can be used to login users through an upstream
+// OpenID Connect provider.
+func (c *Config) Open(logger logrus.FieldLogger) (conn connector.Connector, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	provider, err := oidc.NewProvider(ctx, c.Issuer)
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to get provider: %v", err)
+	}
+	scopes := []string{oidc.ScopeOpenID}
+	if len(c.Scopes) > 0 {
+		scopes = append(scopes, c.Scopes...)
+	} else {
+		scopes = append(scopes, "profile", "email")
+	}
+	clientID := c.ClientID
+	return &googleConnector{
+		redirectURI: c.RedirectURI,
+		oauth2Config: &oauth2.Config{
+			ClientID:     clientID,
+			ClientSecret: c.ClientSecret,
+			Endpoint:     provider.Endpoint(),
+			Scopes:       scopes,
+			RedirectURL:  c.RedirectURI,
+		},
+		verifier: provider.Verifier(
+			&oidc.Config{ClientID: clientID},
+		),
+		logger:        logger,
+		cancel:        cancel,
+		hostedDomains: c.HostedDomains,
+		serviceAccountCred: c.ServiceAccountCred,
+		subject:        c.Subject,
+		groupPrefix:	c.GroupPrefix,
+	}, nil
+}
+var (
+	_ connector.CallbackConnector = (*googleConnector)(nil)
+	_ connector.RefreshConnector  = (*googleConnector)(nil)
+)
+type googleConnector struct {
+	redirectURI   string
+	oauth2Config  *oauth2.Config
+	verifier      *oidc.IDTokenVerifier
+	ctx           context.Context
+	cancel        context.CancelFunc
+	logger        logrus.FieldLogger
+	hostedDomains []string
+	serviceAccountCred  string
```

Comment:
this needs to be a path to a json file instead of the contents of the json file. That way we can store the secret stuff in a kubernetes secret and mount it in the pod like we do with the ssl certs. The file can be read in the initialization logic

Reply:
importing `io/ioutil` allows you to call `contents, err := ioutil.ReadFile(filepath)`